File: src/lib/circuit-breaker.ts
/**
 * Circuit Breaker implementation for agent error handling and fallback strategies
 * Provides resilience patterns for multi-agent system reliability
 */

import { recordAgentError, AgentName } from './telemetry';

export enum CircuitState {
  CLOSED = 'CLOSED',     // Normal operation
  OPEN = 'OPEN',         // Circuit is open, calls fail fast
  HALF_OPEN = 'HALF_OPEN' // Testing if service has recovered
}

export interface CircuitBreakerConfig {
  failureThreshold: number;    // Number of failures before opening
  resetTimeout: number;        // Time to wait before trying half-open (ms)
  halfOpenRequests: number;    // Number of requests to test in half-open
  monitoringWindow: number;    // Time window for failure counting (ms)
}

export interface CircuitBreakerMetrics {
  totalRequests: number;
  successRequests: number;
  failureRequests: number;
  timeouts: number;
  circuitOpenTime?: number;
  lastFailureTime?: number;
  consecutiveFailures: number;
}

export class CircuitBreaker {
  private state: CircuitState = CircuitState.CLOSED;
  private failures: number = 0;
  private lastFailureTime: number = 0;
  private halfOpenCount: number = 0;
  private metrics: CircuitBreakerMetrics;
  private failureWindow: number[] = [];

  constructor(
    private name: string,
    private agentName: AgentName,
    private config: CircuitBreakerConfig
  ) {
    this.metrics = {
      totalRequests: 0,
      successRequests: 0,
      failureRequests: 0,
      timeouts: 0,
      consecutiveFailures: 0
    };
  }

  async execute<T>(
    operation: () => Promise<T>,
    fallback: () => T | Promise<T>,
    timeout?: number
  ): Promise<T> {
    this.metrics.totalRequests++;

    // Fail fast if circuit is open
    if (this.state === CircuitState.OPEN) {
      if (this.shouldAttemptReset()) {
        this.state = CircuitState.HALF_OPEN;
        this.halfOpenCount = 0;
      } else {
        console.log(`Circuit breaker ${this.name} is OPEN - using fallback`);
        return this.executeFallback(fallback);
      }
    }

    try {
      const result = timeout 
        ? await this.executeWithTimeout(operation, timeout)
        : await operation();

      this.onSuccess();
      return result;

    } catch (error: any) {
      this.onFailure(error);

      // In half-open state, fall back immediately on failure
      if (this.state === CircuitState.OPEN) {
        return this.executeFallback(fallback);
      }

      throw error;
    }
  }

  private async executeWithTimeout<T>(
    operation: () => Promise<T>,
    timeoutMs: number
  ): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.metrics.timeouts++;
        reject(new Error(`Operation timed out after ${timeoutMs}ms`));
      }, timeoutMs);

      operation()
        .then(result => {
          clearTimeout(timer);
          resolve(result);
        })
        .catch(error => {
          clearTimeout(timer);
          reject(error);
        });
    });
  }

  private async executeFallback<T>(fallback: () => T | Promise<T>): Promise<T> {
    try {
      return await fallback();
    } catch (fallbackError: any) {
      recordAgentError(this.agentName, 'fallback_execution', fallbackError, {
        circuitBreakerName: this.name,
        circuitState: this.state
      });
      throw new Error(`Both primary operation and fallback failed: ${fallbackError.message}`);
    }
  }

  private onSuccess(): void {
    this.failures = 0;
    this.metrics.successRequests++;
    this.metrics.consecutiveFailures = 0;
    this.cleanFailureWindow();

    if (this.state === CircuitState.HALF_OPEN) {
      this.halfOpenCount++;
      
      if (this.halfOpenCount >= this.config.halfOpenRequests) {
        this.reset();
      }
    }
  }

  private onFailure(error: Error): void {
    this.failures++;
    this.metrics.failureRequests++;
    this.metrics.consecutiveFailures++;
    this.lastFailureTime = Date.now();
    this.failureWindow.push(this.lastFailureTime);

    recordAgentError(this.agentName, 'circuit_breaker_failure', error, {
      circuitBreakerName: this.name,
      consecutiveFailures: this.metrics.consecutiveFailures,
      circuitState: this.state
    });

    this.cleanFailureWindow();

    // Check if we should open the circuit
    if (this.shouldOpenCircuit()) {
      this.openCircuit();
    }
  }

  private shouldOpenCircuit(): boolean {
    // Open if we have too many failures in the monitoring window
    const recentFailures = this.failureWindow.length;
    const consecutiveFailures = this.metrics.consecutiveFailures;

    return (
      consecutiveFailures >= this.config.failureThreshold ||
      recentFailures >= this.config.failureThreshold
    );
  }

  private openCircuit(): void {
    this.state = CircuitState.OPEN;
    this.metrics.circuitOpenTime = Date.now();
    this.halfOpenCount = 0;

    console.warn(`Circuit breaker ${this.name} opened due to failures`, {
      consecutiveFailures: this.metrics.consecutiveFailures,
      recentFailures: this.failureWindow.length,
      agentName: this.agentName
    });
  }

  private shouldAttemptReset(): boolean {
    const timeSinceOpen = Date.now() - (this.metrics.circuitOpenTime || 0);
    return timeSinceOpen >= this.config.resetTimeout;
  }

  private reset(): void {
    this.state = CircuitState.CLOSED;
    this.failures = 0;
    this.halfOpenCount = 0;
    this.failureWindow = [];
    this.metrics.consecutiveFailures = 0;
    delete this.metrics.circuitOpenTime;

    console.info(`Circuit breaker ${this.name} reset to CLOSED state`);
  }

  private cleanFailureWindow(): void {
    const cutoff = Date.now() - this.config.monitoringWindow;
    this.failureWindow = this.failureWindow.filter(time => time > cutoff);
  }

  // Public methods for monitoring
  getState(): CircuitState {
    return this.state;
  }

  getMetrics(): CircuitBreakerMetrics & { 
    state: CircuitState;
    recentFailures: number;
    failureRate: number;
  } {
    const totalInWindow = this.metrics.totalRequests > 0 ? this.metrics.totalRequests : 1;
    
    return {
      ...this.metrics,
      state: this.state,
      recentFailures: this.failureWindow.length,
      failureRate: this.metrics.failureRequests / totalInWindow
    };
  }

  forceOpen(): void {
    this.state = CircuitState.OPEN;
    this.metrics.circuitOpenTime = Date.now();
    console.warn(`Circuit breaker ${this.name} manually forced open`);
  }

  forceClose(): void {
    this.reset();
    console.info(`Circuit breaker ${this.name} manually forced closed`);
  }
}

/**
 * Circuit Breaker Manager for all agents
 */
export class CircuitBreakerManager {
  private breakers = new Map<string, CircuitBreaker>();
  private defaultConfig: CircuitBreakerConfig = {
    failureThreshold: 5,
    resetTimeout: 60000, // 1 minute
    halfOpenRequests: 3,
    monitoringWindow: 300000 // 5 minutes
  };

  getCircuitBreaker(
    agentName: AgentName,
    operation: string,
    config?: Partial<CircuitBreakerConfig>
  ): CircuitBreaker {
    const key = `${agentName}-${operation}`;
    
    if (!this.breakers.has(key)) {
      const breakerConfig = { ...this.defaultConfig, ...config };
      const breaker = new CircuitBreaker(key, agentName, breakerConfig);
      this.breakers.set(key, breaker);
    }

    return this.breakers.get(key)!;
  }

  getAllMetrics(): Map<string, any> {
    const allMetrics = new Map();
    
    for (const [key, breaker] of this.breakers) {
      allMetrics.set(key, breaker.getMetrics());
    }

    return allMetrics;
  }

  getHealthSummary(): {
    healthy: string[];
    degraded: string[];
    failed: string[];
  } {
    const summary = {
      healthy: [] as string[],
      degraded: [] as string[],
      failed: [] as string[]
    };

    for (const [key, breaker] of this.breakers) {
      const metrics = breaker.getMetrics();
      
      if (metrics.state === CircuitState.OPEN) {
        summary.failed.push(key);
      } else if (metrics.state === CircuitState.HALF_OPEN || metrics.failureRate > 0.1) {
        summary.degraded.push(key);
      } else {
        summary.healthy.push(key);
      }
    }

    return summary;
  }

  // Emergency controls
  openAll(): void {
    for (const breaker of this.breakers.values()) {
      breaker.forceOpen();
    }
  }

  closeAll(): void {
    for (const breaker of this.breakers.values()) {
      breaker.forceClose();
    }
  }
}

// Singleton instance
export const circuitBreakerManager = new CircuitBreakerManager();

/**
 * Decorator for automatic circuit breaker protection
 */
export function withCircuitBreaker(
  agentName: AgentName,
  operation: string,
  config?: Partial<CircuitBreakerConfig>
) {
  return function (
    target: any,
    propertyName: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;

    descriptor.value = async function (...args: any[]) {
      const circuitBreaker = circuitBreakerManager.getCircuitBreaker(
        agentName,
        operation,
        config
      );

      // Define fallback function (should be customized per use case)
      const fallback = () => {
        throw new Error(`Service ${agentName} is temporarily unavailable`);
      };

      return circuitBreaker.execute(
        () => originalMethod.apply(this, args),
        fallback,
        config?.resetTimeout
      );
    };

    return descriptor;
  };
}

export default {
  CircuitBreaker,
  CircuitBreakerManager,
  circuitBreakerManager,
  withCircuitBreaker,
  CircuitState
};

File: backends/interview-agents-python/common/circuit_breaker.py
"""
Circuit Breaker implementation for agent error handling and fallback strategies.
Provides resilience patterns for multi-agent system reliability.
Ported from TypeScript implementation in src/lib/circuit-breaker.ts
"""

import asyncio
import time
import logging
from typing import Dict, Any, List, Optional, Callable, TypeVar, Union
from enum import Enum
from dataclasses import dataclass, field

from .config import AgentName
from .telemetry import record_agent_error, agent_metrics

logger = logging.getLogger(__name__)

T = TypeVar('T')


class CircuitState(str, Enum):
    """Circuit breaker states."""
    CLOSED = "CLOSED"      # Normal operation
    OPEN = "OPEN"          # Circuit is open, calls fail fast
    HALF_OPEN = "HALF_OPEN"  # Testing if service has recovered


@dataclass
class CircuitBreakerConfig:
    """Circuit breaker configuration."""
    failure_threshold: int = 5           # Number of failures before opening
    reset_timeout: int = 60000          # Time to wait before trying half-open (ms)
    half_open_requests: int = 3         # Number of requests to test in half-open
    monitoring_window: int = 300000     # Time window for failure counting (ms)


@dataclass
class CircuitBreakerMetrics:
    """Circuit breaker metrics."""
    total_requests: int = 0
    success_requests: int = 0
    failure_requests: int = 0
    timeouts: int = 0
    circuit_open_time: Optional[int] = None
    last_failure_time: Optional[int] = None
    consecutive_failures: int = 0


class CircuitBreaker:
    """Circuit breaker for agent operations - matches TypeScript implementation."""
    
    def __init__(
        self,
        name: str,
        agent_name: AgentName,
        config: CircuitBreakerConfig
    ):
        self.name = name
        self.agent_name = agent_name
        self.config = config
        
        self.state = CircuitState.CLOSED
        self.failures = 0
        self.last_failure_time = 0
        self.half_open_count = 0
        self.failure_window: List[int] = []
        
        self.metrics = CircuitBreakerMetrics()
    
    async def execute(
        self,
        operation: Callable[[], T],
        fallback: Callable[[], Union[T, Callable[[], T]]],
        timeout: Optional[int] = None
    ) -> T:
        """Execute operation with circuit breaker protection."""
        self.metrics.total_requests += 1
        
        # Fail fast if circuit is open
        if self.state == CircuitState.OPEN:
            if self._should_attempt_reset():
                self.state = CircuitState.HALF_OPEN
                self.half_open_count = 0
            else:
                logger.info(f"Circuit breaker {self.name} is OPEN - using fallback")
                return await self._execute_fallback(fallback)
        
        try:
            if timeout:
                result = await self._execute_with_timeout(operation, timeout)
            else:
                # Handle both sync and async operations
                if asyncio.iscoroutinefunction(operation):
                    result = await operation()
                else:
                    result = operation()
            
            self._on_success()
            return result
            
        except Exception as error:
            self._on_failure(error)
            
            # In half-open state, fall back immediately on failure
            if self.state == CircuitState.OPEN:
                return await self._execute_fallback(fallback)
            
            raise error
    
    async def _execute_with_timeout(self, operation: Callable[[], T], timeout_ms: int) -> T:
        """Execute operation with timeout."""
        try:
            if asyncio.iscoroutinefunction(operation):
                result = await asyncio.wait_for(operation(), timeout=timeout_ms / 1000.0)
            else:
                # For sync operations, run in executor with timeout
                loop = asyncio.get_event_loop()
                result = await asyncio.wait_for(
                    loop.run_in_executor(None, operation),
                    timeout=timeout_ms / 1000.0
                )
            return result
        except asyncio.TimeoutError:
            self.metrics.timeouts += 1
            raise Exception(f"Operation timed out after {timeout_ms}ms")
    
    async def _execute_fallback(self, fallback: Callable[[], Union[T, Callable[[], T]]]) -> T:
        """Execute fallback function."""
        try:
            if asyncio.iscoroutinefunction(fallback):
                return await fallback()
            else:
                return fallback()
        except Exception as fallback_error:
            record_agent_error(
                self.agent_name,
                'fallback_execution',
                fallback_error,
                {
                    'circuit_breaker_name': self.name,
                    'circuit_state': self.state.value
                }
            )
            raise Exception(f"Both primary operation and fallback failed: {fallback_error}")
    
    def _on_success(self) -> None:
        """Handle successful operation."""
        self.failures = 0
        self.metrics.success_requests += 1
        self.metrics.consecutive_failures = 0
        self._clean_failure_window()
        
        if self.state == CircuitState.HALF_OPEN:
            self.half_open_count += 1
            
            if self.half_open_count >= self.config.half_open_requests:
                self._reset()
    
    def _on_failure(self, error: Exception) -> None:
        """Handle failed operation."""
        self.failures += 1
        self.metrics.failure_requests += 1
        self.metrics.consecutive_failures += 1
        self.last_failure_time = int(time.time() * 1000)  # Convert to milliseconds
        self.failure_window.append(self.last_failure_time)
        
        record_agent_error(
            self.agent_name,
            'circuit_breaker_failure',
            error,
            {
                'circuit_breaker_name': self.name,
                'consecutive_failures': self.metrics.consecutive_failures,
                'circuit_state': self.state.value
            }
        )
        
        self._clean_failure_window()
        
        # Check if we should open the circuit
        if self._should_open_circuit():
            self._open_circuit()
    
    def _should_open_circuit(self) -> bool:
        """Check if circuit should be opened."""
        recent_failures = len(self.failure_window)
        consecutive_failures = self.metrics.consecutive_failures
        
        return (
            consecutive_failures >= self.config.failure_threshold or
            recent_failures >= self.config.failure_threshold
        )
    
    def _open_circuit(self) -> None:
        """Open the circuit."""
        self.state = CircuitState.OPEN
        self.metrics.circuit_open_time = int(time.time() * 1000)
        self.half_open_count = 0
        
        logger.warning(
            f"Circuit breaker {self.name} opened due to failures",
            extra={
                'consecutive_failures': self.metrics.consecutive_failures,
                'recent_failures': len(self.failure_window),
                'agent_name': self.agent_name.value
            }
        )
    
    def _should_attempt_reset(self) -> bool:
        """Check if circuit should attempt reset."""
        if not self.metrics.circuit_open_time:
            return False
        
        time_since_open = int(time.time() * 1000) - self.metrics.circuit_open_time
        return time_since_open >= self.config.reset_timeout
    
    def _reset(self) -> None:
        """Reset circuit to closed state."""
        self.state = CircuitState.CLOSED
        self.failures = 0
        self.half_open_count = 0
        self.failure_window = []
        self.metrics.consecutive_failures = 0
        self.metrics.circuit_open_time = None
        
        logger.info(f"Circuit breaker {self.name} reset to CLOSED state")
    
    def _clean_failure_window(self) -> None:
        """Clean old failures from window."""
        cutoff = int(time.time() * 1000) - self.config.monitoring_window
        self.failure_window = [t for t in self.failure_window if t > cutoff]
    
    # Public methods for monitoring
    def get_state(self) -> CircuitState:
        """Get current circuit state."""
        return self.state
    
    def get_metrics(self) -> Dict[str, Any]:
        """Get circuit breaker metrics."""
        total_in_window = max(self.metrics.total_requests, 1)
        
        return {
            **self.metrics.__dict__,
            'state': self.state.value,
            'recent_failures': len(self.failure_window),
            'failure_rate': self.metrics.failure_requests / total_in_window
        }
    
    def force_open(self) -> None:
        """Manually force circuit open."""
        self.state = CircuitState.OPEN
        self.metrics.circuit_open_time = int(time.time() * 1000)
        logger.warning(f"Circuit breaker {self.name} manually forced open")
    
    def force_close(self) -> None:
        """Manually force circuit closed."""
        self._reset()
        logger.info(f"Circuit breaker {self.name} manually forced closed")


class CircuitBreakerManager:
    """Circuit Breaker Manager for all agents - matches TypeScript implementation."""
    
    def __init__(self):
        self.breakers: Dict[str, CircuitBreaker] = {}
        self.default_config = CircuitBreakerConfig(
            failure_threshold=5,
            reset_timeout=60000,  # 1 minute
            half_open_requests=3,
            monitoring_window=300000  # 5 minutes
        )
    
    def get_circuit_breaker(
        self,
        agent_name: AgentName,
        operation: str,
        config: Optional[CircuitBreakerConfig] = None
    ) -> CircuitBreaker:
        """Get or create circuit breaker for agent operation."""
        key = f"{agent_name.value}-{operation}"
        
        if key not in self.breakers:
            breaker_config = config or self.default_config
            breaker = CircuitBreaker(key, agent_name, breaker_config)
            self.breakers[key] = breaker
        
        return self.breakers[key]
    
    def get_all_metrics(self) -> Dict[str, Any]:
        """Get metrics for all circuit breakers."""
        return {key: breaker.get_metrics() for key, breaker in self.breakers.items()}
    
    def get_health_summary(self) -> Dict[str, List[str]]:
        """Get health summary of all circuit breakers."""
        summary = {
            'healthy': [],
            'degraded': [],
            'failed': []
        }
        
        for key, breaker in self.breakers.items():
            metrics = breaker.get_metrics()
            
            if metrics['state'] == CircuitState.OPEN.value:
                summary['failed'].append(key)
            elif metrics['state'] == CircuitState.HALF_OPEN.value or metrics['failure_rate'] > 0.1:
                summary['degraded'].append(key)
            else:
                summary['healthy'].append(key)
        
        return summary
    
    # Emergency controls
    def open_all(self) -> None:
        """Force open all circuit breakers."""
        for breaker in self.breakers.values():
            breaker.force_open()
    
    def close_all(self) -> None:
        """Force close all circuit breakers."""
        for breaker in self.breakers.values():
            breaker.force_close()


# Singleton instance
circuit_breaker_manager = CircuitBreakerManager()


def with_circuit_breaker(
    agent_name: AgentName,
    operation: str,
    config: Optional[CircuitBreakerConfig] = None
):
    """Decorator for automatic circuit breaker protection - matches TypeScript."""
    def decorator(func: Callable) -> Callable:
        async def wrapper(*args, **kwargs):
            circuit_breaker = circuit_breaker_manager.get_circuit_breaker(
                agent_name,
                operation,
                config
            )
            
            # Define fallback function (should be customized per use case)
            def fallback():
                raise Exception(f"Service {agent_name.value} is temporarily unavailable")
            
            return await circuit_breaker.execute(
                lambda: func(*args, **kwargs),
                fallback,
                config.reset_timeout if config else None
            )
        
        return wrapper
    return decorator